File: 1115 Print FooBar Alternately.py
class FooBar:
    def __init__(self, n):
        from threading import Semaphore

        self.n = n
        self._foo = Semaphore(1)
        self._bar = Semaphore(0)


    def foo(self, printFoo: 'Callable[[], None]') -> None:
        
        for i in range(self.n):
            self._foo.acquire()
            printFoo()
            self._bar.release()


    def bar(self, printBar: 'Callable[[], None]') -> None:
        
        for i in range(self.n):
            self._bar.acquire()
            printBar()
            self._foo.release()

File: 1115 Print FooBar Alternately.java
import java.util.concurrent.Semaphore;

class FooBar {
    private int n;
    private Semaphore fooSem;
    private Semaphore barSem;

    public FooBar(int n) {
        this.n = n;
        this.fooSem = new Semaphore(1);
        this.barSem = new Semaphore(0);
    }

    public void foo(Runnable printFoo) throws InterruptedException {
        for (int i = 0; i < n; i++) {
            this.fooSem.acquire();
            // printFoo.run() outputs "foo". Do not change or remove this line.
            printFoo.run();
            this.barSem.release();
        }
    }

    public void bar(Runnable printBar) throws InterruptedException {
        for (int i = 0; i < n; i++) {
            this.barSem.acquire();
            // printBar.run() outputs "bar". Do not change or remove this line.
            printBar.run();
            this.fooSem.release();
        }
    }
}
